pkg/certificatesigningrequest/controller.go

```diff
@@ -0,0 +1,168 @@
+	clientkubernetes "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+// ProviderChecks is a function that checks if the CertificateSigningRequest is valid in the provider.
+type ProviderChecks func(context.Context, clientkubernetes.Interface, *x509.CertificateRequest) (bool, error)
+
+// Reconciler is the controller for CertificateSigningRequest.
+type Reconciler struct {
+	kclient        clientkubernetes.Interface
+	providerChecks ProviderChecks
+}
+
+// NewCsrController returns a new CertificateSigningRequest controller.
+func NewCsrController(kclient clientkubernetes.Interface, fn ProviderChecks) *Reconciler {
+	return &Reconciler{
+		kclient:        kclient,
+		providerChecks: fn,
+	}
+}
+
+// Run the CertificateSigningRequest controller.
+//
+//nolint:gocyclo
+func (r *Reconciler) Run(ctx context.Context) {
+	watchTimeoutSeconds := int64(time.Minute * 5)
+
+	for {
+		watcher, err := r.kclient.
+			CertificatesV1().
+			CertificateSigningRequests().
+			Watch(ctx, metav1.ListOptions{
+				Watch:          true,
+				TimeoutSeconds: &watchTimeoutSeconds, // Default timeout: 20 minutes.
+			})
+		if err != nil {
+			klog.Errorf("CertificateSigningRequestReconciler: failed to list CSR resources: %v", err)
+			time.Sleep(10 * time.Second) // Pause for a while before retrying, otherwise we'll spam error logs.
+
+			continue
+		}
+
+		csrWatcher := k8swatch.Filter(watcher, func(in k8swatch.Event) (out k8swatch.Event, keep bool) {
+			if in.Type != k8swatch.Added {
+				return in, false
+			}
+
+			return in, true
+		})
+
+	watch:
+		for {
+			select {
+			case <-ctx.Done():
+				klog.V(4).Infof("CertificateSigningRequestReconciler: context canceled, terminating")
+
+				return
+
+			case event, ok := <-csrWatcher.ResultChan():
+				if !ok {
+					// Server timeout closed the watcher channel, loop again to re-create a new one.
+					klog.V(5).Infof("CertificateSigningRequestReconciler: API server closed watcher channel")
+
+					break watch
+				}
+
+				csr, ok := event.Object.DeepCopyObject().(*certificatesv1.CertificateSigningRequest)
+				if !ok {
+					klog.Errorf("CertificateSigningRequestReconciler: expected event of type *CertificateSigningRequest, got %v",
+						event.Object.GetObjectKind())
+
+					continue
+				}
+
+				valid, err := r.Reconcile(ctx, csr)
+				if err != nil {
+					klog.Errorf("CertificateSigningRequestReconciler: failed to reconcile CSR %s: %v", csr.Name, err)
+
+					continue
+				}
+
+				_, err = r.kclient.CertificatesV1().CertificateSigningRequests().UpdateApproval(ctx, csr.Name, csr, metav1.UpdateOptions{})
+				if err != nil {
+					klog.Errorf("CertificateSigningRequestReconciler: failed to approve/deny CSR %s: %v", csr.Name, err)
+				}
+
+				if !valid {
+					klog.V(3).Infof("CertificateSigningRequestReconciler: has been denied: %s, %+v", csr.Name, err.Error())
+				} else {
+					klog.V(3).Infof("CertificateSigningRequestReconciler: has been approved: %s", csr.Name)
+				}
```

Comment:
```suggestion

if !valid {
					klog.V(3).Infof("CertificateSigningRequestReconciler: has been denied: %s, %+v", csr.Name, err.Error())
					
					continue
				} 
				
			_, err = r.kclient.CertificatesV1().CertificateSigningRequests().UpdateApproval(ctx, csr.Name, csr, metav1.UpdateOptions{})
				if err != nil {
					klog.Errorf("CertificateSigningRequestReconciler: failed to approve/deny CSR %s: %v", csr.Name, err)
				}

					klog.V(3).Infof("CertificateSigningRequestReconciler: has been approved: %s", csr.Name)
```

this is probably more clear

Reply:
Unfortunately, we cannot make `continue` here, we have to update check result of CSR.
We need to run it, after all checks and update our decision. 
`
r.kclient.CertificatesV1().CertificateSigningRequests().UpdateApproval(ctx, csr.Name, csr, metav1.UpdateOptions{})
`


the block
```
if !valid {
					klog.V(3).Infof("CertificateSigningRequestReconciler: has been denied: %s, %+v", csr.Name, err.Error())
				} else {
					klog.V(3).Infof("CertificateSigningRequestReconciler: has been approved: %s", csr.Name)
				}
```
For debug only `klog.V(3)`